Day 22: “Semantics beyond the tag name” / Nathan Knowler

knowler opened the pull request:
*Ready for re-review or merge*

~~TODO before re-review and/or merge:~~

- [x] Introduce landmarks and explain the problem of `<section>` in more detail
- [x] Resolve the confusion prior to Implicit ARIA Semantics section
- [x] Make a better “gibberish” example that contains examples that are addressed throughout the post 
    - Related comments: <https://github.com/matuzo/HTMHell/pull/213#discussion_r2529746540>, <https://github.com/matuzo/HTMHell/pull/213#discussion_r2529745184>
- [x] Handle the `sectionheader`/`sectionfooter` section better
- [x] Address [Elise’s concern about testing](https://github.com/matuzo/HTMHell/pull/213#discussion_r2529779798)

knowler committed:
Add day 22: “Semantics beyond the tag name”

knowler commented on hell/adventcalendar/2025/22/index.md:
```diff
@@ -1,35 +1,350 @@
+
+### Why follow the HTML Standard?
+
+Now, I’m not your god and neither is the HTML Standard. You don’t have to follow it if you don’t want to. If you choose to use HTML elements for how you perceive them, there will more than likely be undesirable consequences that you cannot control. Consequences like assistive technologies not working well with your documents which can also lead to legal consequences if there are accessibility law ...
```
Similar to my comment on lines 95-101, this seems like it also could be patronizing. I am trying to play off the theme of the site (i.e. “hell”).

matuzo replied:
Fine for me.

elisehein replied:
Felt fine to me on first read-through as well, but the thought is perhaps a bit long-winded, and that can also make it feel more preachy than it needs to be?

Suggest to keep the humour but just tighten it a bit:

```suggestion
I’m not your god, and neither is the HTML Standard – but if you choose to use HTML elements for how you perceive them, things break in ways you can’t control. Assistive tech, legal requirements, user expectations... it all hinges on the platform’s actual definitions. Ignore them and, well, your documents might just end up [in hell](https://www.htmhell.dev/).
```

knowler replied:
@elisehein That reads so much better. Thanks! Resolved in [365d460](https://github.com/matuzo/HTMHell/pull/213/commits/365d4600d52b678bb8e4a099bd20b132142648f7)

matuzo commented on hell/adventcalendar/2025/22/index.md:
```diff
@@ -1,35 +1,350 @@
+Something that’s important to understand about semantic HTML is that, while HTML may define all of its elements, not all elements are meaningful for assistive technologies. With that said, you can spend a lot of time writing valid, semantically correct HTML and accomplish absolutely nothing for what makes semantic HTML the most valuable: [making the web for everyone](https://www.w3.org/TR/w3c-vis ...
+
+Beyond the tag name and even HTML semantics, there’s another invisible layer of semantics that we need to understand to write semantic HTML well: implicit ARIA semantics.
+
```
This is confusing. For most people, writing semantic HTML means writing HTML that works well for AT. If there are two types of semantics, why would I need the one that doesn't provide anything for AT? Why does it even exist? What are the other benefits of using semantic elements besides implicit ARIA roles?
You're explaining ARIA semantics in detail, that's fine, but it's not clear how you differentiate between HTML semantics and ARIA semantics, and what exactly HTML semantics do and what we need them for.

elisehein replied:
I think maybe this is just a question of wording/positioning? The way "implicit ARIA semantics" it’s introduced here can read as if HTML semantics and ARIA semantics are separate systems, or two kinds of semantics authors need to choose between.

elisehein replied:
Maybe literally just expanding on the definition of "ARIA semantics" would help – the role, properties/states, accessibility mappings etc _derived from HTML semantics_ (rather than an unrelated other system)

knowler replied:
Thinking out loud (again), I think what I’m trying to accomplish here and with this post is to steer readers not only towards using semantic HTML elements, but semantic HTML elements that are impactful for good accessibility. Perhaps a better problem to point out here is: there are a lot of semantic HTML elements — how do we know which ones to prioritize for good accessibility? Then I can answer with introducing ARIA semantics, focusing specifically on the implicit ones (but call out explicit ones since the reader has a better chance at being familiar with them).

knowler replied:
In https://github.com/matuzo/HTMHell/pull/213/commits/1e45ef1ee142ed4947f10263b8307efcd063c557, I decided to introduce ARIA first, rebutting the idea that one might use ARIA instead of semantic HTML and then using this as an opportunity to introduce semantic HTML’s implicit ARIA which I hope avoids setting them up as competing semantics. Since I’ve done this, I slightly adjusted the title of the next section to reflect the change of where the ARIA definition is, since that section now only focuses on how implicit ARIA works in HTML.

elisehein replied:
Love this direction!

matuzo commented on hell/adventcalendar/2025/22/index.md:
````diff
@@ -1,35 +1,350 @@
+If you’re unfamiliar, <dfn><i>divitis</i></dfn> is a condition where a website uses an unholy amount of `<div>` elements. Some of the worst cases even include `<div>`s used as buttons—_blasphemy!_
 
-Some text.
+```html
+<div class="site-header">
+  <div class="site-title">divs r us</div>
+  <div class="navigation">
+    <div class="link">my favourite divs</div>
+    <div class="link">moor divs</div>
+  </div>
+</div>
+<div class="main">
+  <div class="content">
+    <div class="container">
+      <div class="container-inner">
+        <div class="card">
+          <div class="card-inner">
+            <div class="card-header">
+              <div>i love div</div>
+            </div>
+            <div>all my friends are divs</div>
+            <div class="card-footer">
+              <div class="button">accept div as your lord and saviour</div>
+            </div>
+          </div>
+        </div>
+      </div>
+    </div>
+  </div>
+</div>
+<div class="site-footer">
+  <div class="div-webring">
+    <div class="button">join the divitis webring</div>
+    <div class="button next">next</div>
+    <div class="button prev">previous</div>
+  </div>
+</div>
+```
````
It's fine for me.

elisehein replied:
Agree it's fine, making it a bit OTT is helpful to the point here I think?

matuzo commented on hell/adventcalendar/2025/22/index.md:
```diff
@@ -1,35 +1,350 @@
+
+You’re likely not a child though, so I will tell it to you straight: _this is not cute._
+
+Once you’ve recovered from that devestating realization, we can proceed.
```
A little bit maybe, but imo it's not too much. We'll see what Elise has to say. :)

elisehein replied:
I didn't feel that way when first reading through it, but if you're worried, maybe my comment above about explaining why it's gibberish will make this land softer.

matuzo commented on hell/adventcalendar/2025/22/index.md:
````diff
@@ -1,35 +1,350 @@
+</section>
+```
+
+With that said, please do not go name _all_ of your `<section>` elements! If they were meant to be named, then the browser would do this for you. Often adding more landmarks makes it more difficult to identify what’s important on a page.
````
> If they were meant to be named, then the browser would do this for you.

I don't understand this sentence. The nav or dialog elements are supposed to be named, but the browser still doesn't do it for me.

matuzo replied:
> Often adding more landmarks

You are referring to landmarks, but you're using the term _landmark_ here for the first time.

knowler replied:
I can see how this was confusing. I expanded my reasoning here and formally introduced the term landmark for the reader in https://github.com/matuzo/HTMHell/pull/213/commits/2bd554d4e2c5aca089118a07911d5fba76b6ccde. I might need to do some more tweaking as I think it’s kind of odd that I call out that `<section>` elements have no ARIA semantics when not named, but then immediately advise against adding names for most cases.

knowler replied:
In [eb16470](https://github.com/matuzo/HTMHell/pull/213/commits/eb164708316b72aefe8192fd41de55e6d1af5553), I made some final adjustments to this section. I now focus on the fact that most `<section>` usage is no different than using a `<div>` and their implicit roles are the same. I think that and explaining the `region` role and landmarks makes it more clear why folks shouldn’t name all of their sections.

matuzo commented on hell/adventcalendar/2025/22/index.md:
````diff
@@ -1,35 +1,350 @@
+</body>
+```
+
+As of writing, this change has only been implemented in Chromium and WebKit-based browsers. [Gecko has yet to implement them](https://bugzilla.mozilla.org/show_bug.cgi?id=1893684). As these are new roles, the support from assistive technologies will likely be non-existent or quite limited.
````
It's worth noting that, before this addition, the implicit role of the header and footer outside the body context was/is _generic_.

knowler replied:
[db2fce2](https://github.com/matuzo/HTMHell/pull/213/commits/db2fce204e7bb9ab1b38fc1071f5392854601e6b) Ok, so I think I improved this section a fair bit. Now, I do explicit mention that they were `generic` prior.

matuzo commented on hell/adventcalendar/2025/22/index.md:
```diff
@@ -1,35 +1,350 @@
+There is a terrible epidemic that plagues the web: <i>divitis</i>.
 
-## Heading
+If you’re unfamiliar, <dfn><i>divitis</i></dfn> is a condition where a website uses an unholy amount of `<div>` elements. Some of the worst cases even include `<div>`s used as buttons—_blasphemy!_
```
> `<div>`s used as buttons
😲

elisehein replied:
🥴

matuzo commented on hell/adventcalendar/2025/22/index.md:
````diff
@@ -1,35 +1,350 @@
+</body>
+```
+
+As of writing, this change has only been implemented in Chromium and WebKit-based browsers. [Gecko has yet to implement them](https://bugzilla.mozilla.org/show_bug.cgi?id=1893684). As these are new roles, the support from assistive technologies will likely be non-existent or quite limited.
````
I'm not sure. As far as I know, they're still debating about whether and how to implement these roles.

matuzo reviewed (commented):
Great post, thank you!

elisehein commented on hell/adventcalendar/2025/22/index.md:
````diff
@@ -1,35 +1,350 @@
+</div>
+```
+
+To avoid <i>divitis</i>, authors are often encouraged to use <i>semantic HTML</i>. As it turns out, HTML has over a hundred elements that are not named `div` or `span`. Semantic HTML elements describe what their content is.
````
Take this as very much of a side note, but in my experience an equally (and perhaps even bigger) cause of divitis is bloat and unnecessary wrappers (often due to insufficient CSS knowledge). Even if authors suddenly used all the right semantic elements, if you squint, the output would still be mostly divs. 

Up to you if worth mentioning (even just replacing with "_One of the ways_ of avoiding..."), just thought I'd note!

knowler replied:
I didn’t end up getting to this and probably won’t given the time, but I do agree!

elisehein commented on hell/adventcalendar/2025/22/index.md:
```diff
@@ -1,35 +1,350 @@
+
+To avoid <i>divitis</i>, authors are often encouraged to use <i>semantic HTML</i>. As it turns out, HTML has over a hundred elements that are not named `div` or `span`. Semantic HTML elements describe what their content is.
+
+Inspired by this newfound responsibility, authors open the lexicon of HTML elements and begin to compose… _absolute gibberish._
```
It's fairly obvious to a certain audience, but since you don't immediately go on to explain why the given example is gibberish, it may leave less knowledgeable readers feeling lost 🥹 Maybe just as a kindness (and also to improve signposting), something along the lines of "(we’ll unpack why this structure doesn’t mean what the author thinks it means in a moment)"...  just a small cue so less experienced readers don’t feel talked down to.

Or, after the example:

> While it can look deceptively semantic at first (no divs!), in this case it doesn't mean what the author thinks it means.

knowler replied:
Thinking out loud (and leaving this as a note for myself mostly), I should make better use of this example throughout the post like you suggest elsewhere. I was a bit lazy with including this example in the first place (i.e. I couldn’t think of anything on my own, so I went through the site to find one).

knowler replied:
Should be resolved in [ffaa2dd](https://github.com/matuzo/HTMHell/pull/213/commits/ffaa2dd8aebcad201afa5fc38b9a49e92a8e66ee)

elisehein replied:
That, and the new example mirroring the divitis one with the internal monologue is a really clever way to make it obvious 😆 love it

elisehein commented on hell/adventcalendar/2025/22/index.md:
````diff
@@ -1,35 +1,350 @@
+</section>
+```
+
+To be fair, this is to be expected of anyone learning a new language, however, a problem on the web is that authors delude themselves into thinking that they are “writing semantic HTML,” therefore gaining all of the benefits that are understood to come with that like accessibility, and they never advance past that stage.
````
Just a pause to shorten the sentences:

```suggestion
To be fair, this is to be expected of anyone learning a new language. However, a problem on the web is that authors delude themselves into thinking that they are “writing semantic HTML,” therefore gaining all of the benefits that are understood to come with that like accessibility, and they never advance past that stage.
```

knowler replied:
Resolved in [365d460](https://github.com/matuzo/HTMHell/pull/213/commits/365d4600d52b678bb8e4a099bd20b132142648f7)

elisehein commented on hell/adventcalendar/2025/22/index.md:
```diff
@@ -1,35 +1,350 @@
+
+> A <dfn>hierarchically correct `main` element</dfn> is one whose ancestor elements are limited to `html`, `body`, `div`, `form` without an accessible name, and autonomous custom elements. Each `main` element must be a hierarchically correct `main` element.
+
+From these descriptions, we can gather that using the `<main>` element for the main content of a card component that’s meant to be used many times throughout a document would be incorrect:
```
Maybe it'd feel more coherent if instead of giving a new example, you could use this as a chance to reference back to the first gibberish example and explain why that doesn't work?

knowler replied:
In [6389673](https://github.com/matuzo/HTMHell/pull/213/commits/6389673db79283331e16516f0500880bb4775615), I changed the original gibberish example, then in [91dde92](https://github.com/matuzo/HTMHell/pull/213/commits/91dde925b63119d889fd9d7873832c887efc8f45) I use a truncated/focused version of that to point out the issue.

elisehein commented on hell/adventcalendar/2025/22/index.md:
```diff
@@ -1,35 +1,350 @@
+
+You may be familiar with [ARIA (Accessible Rich Internet Applications)][aria]. ARIA is a set of roles and attributes that can be used in HTML to add accessibility related semantics to web content.
+
+[The first rule of ARIA usage](https://www.w3.org/TR/using-aria/#firstrule) is: don’t use ARIA if you don’t need to. Why? Because many semantic HTML elements and attributes have what’s called implicit ARIA semantics—as well as the built-in functionality necessary for those semantics.
```
I get the point here, but the phrasing threw me for a moment because just two paragraphs ago you were saying it's NOT enough to just use the right tag names.

knowler replied:
I believe my reworking of the prior section resolves this concern: [1e45ef1](https://github.com/matuzo/HTMHell/pull/213/commits/1e45ef1ee142ed4947f10263b8307efcd063c557)

elisehein commented on hell/adventcalendar/2025/22/index.md:
````diff
@@ -1,35 +1,350 @@
+</section>
+```
+
+With that said, please do not go name _all_ of your `<section>` elements! If they were meant to be named, then the browser would do this for you. Often adding more landmarks makes it more difficult to identify what’s important on a page.
````
I think this lacks some explanation – it first says they're semantically insignificant without a name, but then says not to name them.

knowler replied:
As noted in https://github.com/matuzo/HTMHell/pull/213#discussion_r2532933128, I’ve begun trying to improve this section with an expanded explanation. I will try to further address this concern though.

elisehein commented on hell/adventcalendar/2025/22/index.md:
```diff
@@ -1,35 +1,350 @@
+
+Something that’s important to understand about semantic HTML is that, while HTML may define all of its elements, not all elements are meaningful for assistive technologies. With that said, you can spend a lot of time writing valid, semantically correct HTML and accomplish absolutely nothing for what makes semantic HTML the most valuable: [making the web for everyone](https://www.w3.org/TR/w3c-vis ...
+
+Beyond the tag name and even HTML semantics, there’s another invisible layer of semantics that we need to understand to write semantic HTML well: implicit ARIA semantics.
```
Unsure if this is the best place for it, but for readers less familiar with the topic, I think calling out that there are also _explicit_ semantics (eg `role="navigation"`), and that they always override the implicit role, would help complete the picture.

knowler replied:
I think the changes I made in [1e45ef1](https://github.com/matuzo/HTMHell/pull/213/commits/1e45ef1ee142ed4947f10263b8307efcd063c557) mostly resolve your concern here since it front-loads introducing ARIA as a concept through the use of explicit attributes. It doesn’t mention anything about explicit ARIA overrides implicit ARIA. If time permits, I might try to work that in somewhere, it could even be a highlighted note.

elisehein commented on hell/adventcalendar/2025/22/index.md:
```diff
@@ -1,35 +1,350 @@
+
+When you create custom elements, you become responsible for communicating their purpose to other authors. This should also include any implicit semantics you’ve set, as well as, what you consider good authoring practice to be when using them.
+
+I highly recommend using HTML itself as a model for how you build custom elements. In many ways, that might mean thinking less like an author and more like an implementor or a specification writer. “How do I create elements for others to use?” Implicit semantics are an important part of that because they allow you to [put other authors first](https://www.w3.org/TR/design-principles/#priority-of-c ...
```
Love this part

elisehein commented on hell/adventcalendar/2025/22/index.md:
```diff
@@ -1,35 +1,350 @@
+
+Testing might look like checking what ARIA roles and properties are computed for elements in a browser’s developer tools. It also might look like actually using assistive technologies such as screen readers or voice control. It is very helpful to understand how assistive technology users experience various elements.
+
+Testing helps us understand what is valuable and what we should focus on. I can spend a whole lot of time trying to perfectly mark up all of my words using [text-level semantics][html-text-level-semantics], but then realize through testing that I’ve delivered little to no semantic value to assistive technology users.
```
100% agree, but in my experience, testing with real AT is considered a huge barrier to most devs. They're unfamiliar territory if you don't use AT yourself, and often feel like they're gonna be really tricky to master. The reality is that very few devs ever try it. 

Not sure where I'm going with this tbh, but it's just a thought I found myself having as I was reading this bit – kinda like, huh, well _that's_ wishful thinking...

Maybe a little note of encouragement or links to resources (eg Sara Soueidan's Practical Accessibility course?) would make this feel more approachable?

knowler replied:
This was a good callout. I added some links to how to find the accessibility info in developer tools and added a couple of course recommendations in [08191b0](https://github.com/matuzo/HTMHell/pull/213/commits/08191b0683115716b5e33e8fb06937f92ff6ac84).

elisehein commented on hell/adventcalendar/2025/22/index.md:
```diff
@@ -1,35 +1,350 @@
+
+We’ve learned that:
+
+1. HTML tag names are not a reliable way of understanding what an element is for.
+2. The HTML standard defines what semantic HTML elements are for.
+3. Some semantic HTML elements also have implicit ARIA semantics, and in some cases that depends on what attributes are set, if the element has an accessible name, the context the element is used in, and even a combination of these things.
+4. Custom elements can have implicit ARIA semantics too.
+5. Don’t take anyone’s word for it: test it yourself.
+
```
Nice 🥇

elisehein reviewed (commented):
Hi Nathan 👋 I'm Elise, nice to meet you.

Lovely article, really nicely written and well structured!

I left a few suggestions and a couple of editorial comments – please ignore at will.

knowler committed:
Apply suggestions from review

Co-authored-by: Elise Hein <elisehein@gmail.com>

knowler committed:
Expand on `<section>` naming warning + define landmarks

matuzo commented:
@knowler how's it going? Is the article ready, can I merge it?

knowler commented:
@matuzo Going well. Not ready quite yet. I’ll finish this within the next few days.

Updates:

- November 19 at 01:00:00 UTC-6: made a fair bit of progress on the most major issue. I’ll continue tomorrow (well, later today now 😅) on the remain issues which I’ve detailed in the PR description.
- November 19 at 23:45:00 UTC-6: I’m done most of the larger changes. There are a couple of outstanding issues I want to resolve (again, noted in the PR description) which I’ll do tomorrow ~during the day~ in the evening.

knowler committed:
Correct a few typos

knowler committed:
Use em dash without surrounding spaces for consistency

knowler committed:
Re-work transition into “Implicit ARIA semantics” section

knowler committed:
Correct grammar

knowler committed:
Use more relevant gibberish example

- Has a card component (using a `<my-card>` custom element since people
  think their custom elements are semantic just because of the tag name)
- Includes an example of more than one `<main>` being used
- `<section>` is being used like a `<div>`
- No classes used, because people get dogmatic about that.
- Includes a “link button” which believe it not is something that exists
  in a particular “classless” CSS library.
- Multiple h1 elements then with an h2 used as a subtitle.
- The `<header>` and `<footer>` elements are reckless: globally they’d
  create duplicate banner/contentinfo landmarks, locally they’d either
  be generic or be duplicates scoped to whatever sectioning content both
  cards are in.
- A `<nav>` wrapping the link button for good measure.

knowler committed:
Use a truncated version of the gibberish example for “main”

knowler committed:
Make `<section>` advice less contradicting

knowler committed:
Add some resources and course recommendations to testing section

knowler committed:
Cushion gibberish

knowler committed:
Improve context section

knowler committed:
Minor edits

knowler commented:
Thank you so much for the reviews @matuzo and @elisehein! I feel like this piece is way better because of your suggestions and feedback. I’m done with my changes now. Let me know if you spot anything that needs any adjustment or doesn’t make sense.

elisehein commented on hell/adventcalendar/2025/22/index.md:
```diff
@@ -360,14 +360,16 @@ When you create custom elements, you become responsible for communicating their
 
 Testing helps us understand what is valuable and what we should focus on. I can spend a whole lot of time trying to perfectly mark up all of my words using [text-level semantics][html-text-level-semantics], but then realize through testing that I’ve delivered little to no semantic value to assistive technology users.
 
+Courses such as [Sara Soueidan’s Practical Accessibility](https://practical-accessibility.today) or [Marcy Sutton’s Testing Accessibility](https://testingaccessibility.com) are great resources for those wanting to get started with testing various assistive technologies.
```
❤️

elisehein commented on hell/adventcalendar/2025/22/index.md:
```diff
@@ -1,35 +1,413 @@
+    </section>
+  </header>
+  <main>
+    <section>whenever I don’t know what element to choose, section is there for me. section is my best friend forever.</section>
```
🤣

elisehein commented on hell/adventcalendar/2025/22/index.md:
````diff
@@ -1,35 +1,413 @@
+
+[Accessible Rich Internet Applications][aria], more commonly known by the acronym ARIA, is a set of roles and attributes that can be used in HTML to add accessibility related semantics to web content.
+
+It can be tempting to think that we don’t need semantic HTML, instead we can just sprinkle our <i>divitis</i> with <em>holy</em> ARIA roles and attributes. Well, while this can improve the accessibility of that hoard of elements, we’re likely breaking [the first rule of ARIA usage](https://www.w3.org/TR/using-aria/#firstrule): don’t use ARIA if you don’t need to.
+
+```html
+<div role=button>accept div as your lord and saviour</div>
+```
+
+Many semantic HTML elements have what’s called <i>implicit ARIA semantics</i>. These are ARIA roles and properties that the browser has mapped for us. Instead of needing to explicitly write out the ARIA semantics for an element each time you use it, we can use semantic HTML elements as a shorthand for these.
+
+```html
+<button>This machine kills fascists</button>
+```
+
+Beyond simply being a shorthand for a set of ARIA semantics, built-in HTML elements also include the necessary functionality and presentation that would be expected of an element bearing such semantics. In many cases, this means that you don’t need to write any JavaScript for your HTML to work as expected. Using JavaScript in lieu of built-in HTML functionality is wasteful and error prone, degrad ...
+
+Now that we understand the folly of ignoring the first rule of ARIA usage, let’s turn our attention to <i>implicit ARIA semantics</i> and how they work in semantic HTML.
````
Nice 👌

elisehein commented on hell/adventcalendar/2025/22/index.md:
```diff
@@ -1,35 +1,413 @@
+
+The problem here is that everything is a landmark, then nothing is a landmark.
+
+It can be easy to think that labelling your `<section>` elements filling a requirement for using them, but that is not the case here. Both the meaning of the tag name and the HTML definition of a `<section>` element is way more generic than the `region` ARIA role. Before you label your `<section>`  elements ensure they make sense as regions, otherwise, you’ll likely be making it more difficult to ...
```
I think a word has gone missing here

```suggestion
It can be easy to think that labelling your `<section>` elements satisfies the requirement for using them, but that is not the case here. Both the meaning of the tag name and the HTML definition of a `<section>` element is way more generic than the `region` ARIA role. Before you label your `<section>`  elements ensure they make sense as regions, otherwise, you’ll likely be making it more difficult to find what’s important on a page. If you’re already using [a good heading structure](https://tetralogical.com/blog/2022/02/28/headings/) throughout your document, that is good enough for what you’re likely trying to do.
```

elisehein replied:
This whole section is much more accessible now!

Also, as a complete aside, irrelevant to this post, but genuinely interested in what you think – what are your thoughts on the DX vs UX aspects of using HTML tagnames like `<section>` that don't make a difference re literal assistive tech usability, but just make it ✨ nicer ✨ to the devs having to iterate on the code?

As a developer working in a team, I often use HTML elements with an implicit `generic` role literally just to improve readability of code (come to think, I'd prob do this when working on a solo project too). In a real project, the HTML for a single view might be scattered across very many files, so any hints at the structure and intended nesting are useful.

I wonder if there can be any risks to this? I guess it comes down to being mindful of intent.

knowler replied:
@elisehein I had this question in the back of mind while writing this and it’s something I’ve changed my mind about in the last year or so. Now, I tend to agree.

When I wrote [the precursor to this article](https://knowler.dev/blog/context-matters-for-semantic-html) a couple of years ago, with regard to `<header>` and `<footer>` elements, I thought that using them outside of contexts where they had (ARIA) semantic value was confusing and it unnecessarily complicated selecting them with CSS (i.e. if you only have one of each of these, then you don’t need to invent a class name or use a selector with more specificity).

At the same time, I had already been using elements with valid custom element tag names instead (e.g. rather than `<article><header>`, I’d use `<article><article-header>`), but that creates a styling burden in another way since the initial `display` value is `inline`—something that is particularly noticeable on [CSS Naked Day](https://css-naked-day.org/)—so I largely stopped doing that for my personal projects, but I’m not opposed to it (this approach is safer for replacing `span` elements).

Now, I think that I’d probably just use elements according to their HTML definition even if they have no ARIA semantic value in certain contexts, unless there was a reason that it was problematic. I definitely use `<section>` a lot more in cases that it doesn’t have any implicit ARIA semantic value—I can only think of two or three cases where I’ve ever associated a label to get `role=region` out of it.

> I wonder if there can be any risks to this? I guess it comes down to being mindful of intent.

I think the whole `sectionheader`/`sectionfooter` change for `<header>`/`<footer>` demonstrates what could happen: if someone had good authoring practices, then they’ll benefit from the semantic upgrade, but if they weren’t used correctly, there could be some unexpected consequence. I presume those making the decisions to add the roles noticed that authoring practices were generally good (the elements are named pretty well).

Maybe another risk: someone who is inexperienced in accessibility comes along and starts to make changes that activate the implicit roles for these elements (or even adds them explicitly). It’s easy to accidentally read [MDN’s “Implicit ARIA role” for `<section>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/section#technical_summary) and conclude “we’re not using `<section>` correctly unless we add an accessible name.”

I do think the risks are generally worth it. Ideally, HTML/HTML-AAM shouldn’t be making breaking changes, so using HTML according to how it defines itself should be safe.

elisehein replied:
Thanks for taking the time to share your thoughts, interesting stuff!

knowler committed:
Fix wording

Co-authored-by: Elise Hein <elisehein@gmail.com>

elisehein commented:
✅ ✅ ✅ , but I think leave the actual merging to @matuzo!

matuzo commented:
Perfect, great job, thank you!
